### DISCUSSION TITLE
Ipopt with OneAPI, HSL problem

### DISCUSSION THREAD

**Discussion by Leersmaekers** [coin-or/Ipopt#656] [Q&A] [answered]
I'm having some troubles installing Ipopt with with HSL and OneAPI (MKL). If I configure&install Ipopt without HSL, it all seems to work as expected. The OneAPI MKL Lapack libraries are found. However, if I first try to configure HSL with MKL Lapack, I reveive an error stating it cannot be linked with the user-specified Lapack. The line I used to link Lapack is the following:

`--with-lapack-lflags="-L${MKLROOT}/lib/intel64 -Wl,--no-as-needed -lmkl_intel_lp64 -lmkl_sequential -lmkl_core -lm"`

Does anyone have some suggestion to fix this issue?
(I did search online for a couple of hours, but nothing seems to fix my issue)

[config.log](https://github.com/coin-or/Ipopt/files/11171136/config.log)

**Comment by svigerske** [comment]
It's a bit odd that if it worked for Ipopt, it didn't for HSL. Can you attach also the config.log from the build of Ipopt without HSL?

It would be interesting to see how the Lapack test in Ipopt is different from the one in HSL (the one failing in config.log:725).

**Reply by tkralphs** [reply]
`MKLROOT` is set to `C:\PROGRA~2\Intel\oneAPI\mkl\2023.0.0`, which is a Windows-style path, whereas you are using Posix paths for everything else. I don't really see how this can work. Maybe try setting `MKLROOT` to a Posix path (or else just specify the path explicitly).

**Reply by Leersmaekers** [reply]
This is the log file that succeeded:

[config.log](https://github.com/coin-or/Ipopt/files/11171477/config.log)

**Comment by svigerske** [comment, marked as answer]
Looks like you use GCC when building HSL, but MSVC when building Ipopt.

For HSL, this one fails:
```
configure:21359: gcc -o conftest.exe -O2 -DNDEBUG    conftest.c -LC:\PROGRA~2\Intel\oneAPI\mkl\2023.0.0/lib/intel64 -Wl,--no-as-needed -lmkl_lapack95_lp64 -lmkl_intel_lp64 -lmkl_sequential -lmkl_core -lm  >&5
```

For Ipopt, this one succeeds:
```
configure:22983: checking for LAPACK
configure:23343: /home/LaurentK/Ipopt/compile cl -o conftest.exe -nologo -DNDEBUG -O2 -MD    conftest.c /c/PROGRA~2/Intel/oneAPI/mkl/20230~1.0/lib/intel64/mkl_intel_lp64_dll.lib /c/PROGRA~2/Intel/oneAPI/mkl/20230~1.0/lib/intel64/mkl_sequential_dll.lib /c/PROGRA~2/Intel/oneAPI/mkl/20230~1.0/lib/intel64/mkl_core_dll.lib  >&5
```

Try adding `--enable-msvc` also for the HSL build.

**Reply by svigerske** [reply]
Also, as for Ipopt, you won't have to specify the `--with-lapack-lflags` when building HSL with `--enable-msvc`. Your MS compiler seems to find the libs automatically. Your GCC doesn't.

**Reply by Leersmaekers** [reply]
Okay, the `--enable-msvc` made it work, thanks a lot!

Just one additional question: is the usage of HSL considered as an advantage in comparison to what's already available within MKL?

**Reply by svigerske** [reply]
It can happen that the HSL linear solvers give you better performance than Pardiso from MKL.

**Reply by Leersmaekers** [reply]
Yesterday the HSL configure worked fine, but configuring Ipopt with HSL now fails (log attached). 

`LINK : fatal error LNK1181: cannot open input file 'coinhsl.lib'`

I can see a libs folder in ThirdParty-HSL\.libs, containing the following files:

- coinhsl.dll.exp
- coinhsl.dll.lib
- coinhsl.exp
- coinhsl-2.dll
- coinhsl-2.dll.exp
- libcoinhsl.la
- libcoinhsl.lai
- metis_adapter.obj

[config.log](https://github.com/coin-or/Ipopt/files/11180917/config.log)

**Reply by svigerske** [reply]
The `make install` in ThirdParty-HSL should have installed the HSL lib together with a `$prefix/lib/pkgconfig/coinhsl.pc` and Ipopt's configure should pick the flags on how to link against HSL from the `.pc` file.
In particular, the coinhsl.pc should have a line `Libs: -L${libdir} -lcoinhsl` and pkg-config should resolve the `$libdir` by some path. But somehow the whole `-L` flag is missing when Ipopt tries to link against the HSL lib.
Can you check whether `$prefix/lib/pkgconfig/coinhsl.pc` was installed and post its content?

**Reply by tkralphs** [reply]
I guess that `$prefix/lib/pkgconfig/coinhsl.pc` also has to be in the `PKG_CONFIG_PATH`. Is Ipopt being installed into the same prefix? Then this should happen automatically, but you can check whether it's working by just doing 
```
PKG_CONFIG_PATH=$prefix/lib/pkgconfig pkg-config --libs coinhsl
```

**Reply by Leersmaekers** [reply]
Yes, it created a `coinhsl.pc` file (see attachment). I modified the extension in the attachment, since a .pc was somehow not allowed.

[coinhsl.txt](https://github.com/coin-or/Ipopt/files/11184986/coinhsl.txt)

**Reply by Leersmaekers** [reply]
My file structure now looks as follows:
home/LaurentK
  |- Ipopt
  |- ThirdParty-HSL (I used this folder for configure/make/install)
  |- ThirdParty
-----> ASL
-----> HSL (I didn't use this folder for configure/make/install)
-----> Mumps

**Reply by svigerske** [reply]
Hmm, strange. I don't really know what is going wrong. (And I actually fear that configure would still fail if the `-L` flag were not lost.)

Maybe just try running Ipopt's configure with
```
--with-hsl-lflags=/usr/lib/coinhsl.dll.lib --with-hsl-cflags=-I/usr/include/coin-or/hsl
```

**Reply by Leersmaekers** [reply]
The line you suggested works perfectly fine. It's indeed strange that the other method didn't work.

After performing `make install`, are there `include` and `lib` folders created that can be used in Visual Studio? I now use the folders as provided with the Windows binaries.

**Reply by svigerske** [reply]
Yes, as far as I see, you get header files in `/usr/include/coin-or`, `.lib`'s in `/usr/lib`, and DLLs in `/usr/bin`.

**Reply by Leersmaekers** [reply]
They are indeed installed in there. I was searching for an `include` folder, but it's indeed one level lower in `coin-or`. Thanks a lot for the quick answers!
